internal/brokers/examplebroker/broker.go

```diff
@@ -220,50 +331,61 @@ func (b *Broker) GetAuthenticationModes(ctx context.Context, sessionID string, s
 
 		case "webview":
 			// This broker does not support webview
+
+		case "newpassword":
+			if layout["entry"] == "" {
+				break
+			}
+			allModes["mandatoryreset"] = map[string]string{
+				"selection_label": "Password reset",
+				"ui": mapToJSON(map[string]string{
+					"type":  "newpassword",
+					"label": "Enter your new password",
+					"entry": "chars_password",
+				}),
+			}
+
+			if layout["skip-button"] != "" {
+				allModes["optionalreset"] = map[string]string{
+					"selection_label": "Password reset",
+					"ui": mapToJSON(map[string]string{
+						"type":        "newpassword",
+						"label":       "Enter your new password",
+						"entry":       "chars_password",
+						"skip-button": "Skip",
+					}),
+				}
+			}
 		}
 	}
 
-	// Sort in preference order. We want by default password as first and potentially last selection too.
-	b.userLastSelectedModeMu.Lock()
-	lastSelection := b.userLastSelectedMode[sessionInfo.username]
-	if _, exists := allModes[lastSelection]; !exists {
-		lastSelection = ""
-	}
-	b.userLastSelectedModeMu.Unlock()
+	return allModes
+}
 
-	var allModeIDs []string
-	for n := range allModes {
-		if n == "password" || n == lastSelection {
-			continue
-		}
-		allModeIDs = append(allModeIDs, n)
+func getMfaModes(info sessionInfo, supportedModes map[string]map[string]string) map[string]map[string]string {
+	mfaModes := make(map[string]map[string]string)
+
+	if authMode, exists := supportedModes["phoneack1"]; exists && info.selectedMode != "phoneack1" {
+		mfaModes["phoneack1"] = authMode
 	}
-	sort.Strings(allModeIDs)
-	if lastSelection != "" && lastSelection != "password" {
-		allModeIDs = append([]string{lastSelection, "password"}, allModeIDs...)
-	} else {
-		allModeIDs = append([]string{"password"}, allModeIDs...)
+	if authMode, exists := supportedModes["totp_with_button"]; exists && info.selectedMode != "totp_with_button" {
+		mfaModes["totp_with_button"] = authMode
 	}
-
-	for _, id := range allModeIDs {
-		authMode := allModes[id]
-		authenticationModes = append(authenticationModes, map[string]string{
-			"id":    id,
-			"label": authMode["selection_label"],
-		})
+	if authMode, exists := supportedModes["fidodevice1"]; exists && info.selectedMode != "fidodevice1" {
+		mfaModes["fidodevice1"] = authMode
 	}
-	sessionInfo.allModes = allModes
+	return mfaModes
+}
 
-	// Checks if the session was ended in the meantime, otherwise we would just accidentally create a new one.
-	if _, err := b.sessionInfo(sessionID); err != nil {
-		return nil, err
+func getPasswdResetModes(info sessionInfo, supportedModes map[string]map[string]string) map[string]map[string]string {
+	passwdResetModes := make(map[string]map[string]string)
+	if authMode, exists := supportedModes["mandatoryreset"]; info.needsPasswdReset && exists {
```

Comment:
I think this could be simplified too.

```go
if !info.needsPasswdReset && !info.canPasswdReset {
    return passwdResetModes
}

desiredMode = "mandatoryreset"
if info.canPasswdReset {
  desiredMode = "optionalreset"
}
<do the autMode collection> check and assign to passwdResetModes that you return.
 ```